services/webtheme/webtheme.go

```diff
@@ -12,63 +13,146 @@ import (
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/public"
 	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/util"
 )
 
 var (
-	availableThemes    []string
-	availableThemesSet container.Set[string]
-	themeOnce          sync.Once
+	availableThemes             []*ThemeMetaInfo
+	availableThemeInternalNames container.Set[string]
+	themeOnce                   sync.Once
 )
 
+const (
+	fileNamePrefix = "theme-"
+	fileNameSuffix = ".css"
+)
+
+type ThemeMetaInfo struct {
+	FileName     string
+	InternalName string
+	DisplayName  string
+}
+
+func parseThemeMetaInfoToMap(cssContent string) map[string]string {
+	/*
+		The theme meta info is stored in the CSS file's variables of `gitea-theme-meta-info` element,
+		which is a privately defined and is only used by backend to extract the meta info.
+		Not using ":root" because it is difficult to parse various ":root" blocks when importing other files,
+		it is difficult to control the overriding, and it's difficult to avoid user's customized overridden styles.
+	*/
+	metaInfoContent := cssContent
+	if pos := strings.LastIndex(metaInfoContent, "gitea-theme-meta-info"); pos >= 0 {
+		metaInfoContent = metaInfoContent[pos:]
+	}
+
+	reMetaInfoItem := `
+(
+\s*(--[-\w]+)
+\s*:
+\s*("(\\"|[^"])*")
```

Comment:
CSS strings can have single or no quotes, add tests for those cases.

Reply:
Done in fdfd4400615afc19d4d42524aeac1f8dffa6ed2d